storm-server/src/main/java/org/apache/storm/metricstore/rocksdb/RocksDbMetricsWriter.java

```diff
@@ -0,0 +1,306 @@
+            stringMetadata.update(metricTimestamp, type);
+            return stringMetadata.getStringId();
+        }
+
+        // attempt to find the string in the database
+        stringMetadata = store.rocksDbGetStringMetadata(type, s);
+        if (stringMetadata != null) {
+            // update to the latest timestamp and add to the string cache
+            stringMetadata.update(metricTimestamp, type);
+            stringMetadataCache.put(s, stringMetadata, false);
+            return stringMetadata.getStringId();
+        }
+
+        // string does not exist, create using an unique string id and add to cache
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(type + "." + s + " does not exist in cache or database");
+        }
+        int stringId = getUniqueMetadataStringId();
+        stringMetadata = new StringMetadata(type, stringId, metricTimestamp);
+        stringMetadataCache.put(s, stringMetadata, true);
+
+        return stringMetadata.getStringId();
+    }
+
+    // get a currently unused unique string id
+    private int getUniqueMetadataStringId() throws MetricException {
+        generateUniqueStringIds();
+        int id = unusedIds.iterator().next();
+        unusedIds.remove(id);
+        return  id;
+    }
+
+    // guarantees a list of unused string Ids exists.  Once the list is empty, creates a new list
+    // by generating a list of random numbers and removing the ones that already are in use.
+    private void generateUniqueStringIds() throws MetricException {
+        int attempts = 0;
+        while (unusedIds.isEmpty()) {
+            attempts++;
+            if (attempts > 100) {
+                String message = "Failed to generate unique ids";
+                LOG.error(message);
+                throw new MetricException(message);
+            }
+            for (int i = 0; i < 600; i++) {
+                int n = ThreadLocalRandom.current().nextInt();
+                if (n == RocksDbStore.INVALID_METADATA_STRING_ID) {
+                    continue;
+                }
+                // remove any entries in the cache
+                if (stringMetadataCache.contains(n)) {
+                    continue;
+                }
+                unusedIds.add(n);
+            }
+            // now scan all metadata and remove any matching string Ids from this list
+            RocksDbKey firstPrefix = RocksDbKey.getPrefix(KeyType.METADATA_STRING_START);
+            RocksDbKey lastPrefix = RocksDbKey.getPrefix(KeyType.METADATA_STRING_END);
+            store.scanRange(firstPrefix, lastPrefix, (key, value) -> {
+                unusedIds.remove(key.getMetadataStringId());
+                return true; // process all metadata
+            });
+        }
+    }
+
+    // writes multiple metric values into the database as a batch operation
+    private void processBatchInsert(Map<RocksDbKey, RocksDbValue> batchMap) throws MetricException {
+        try (WriteBatch writeBatch = new WriteBatch()) {
```

Comment:
Because of the thread safety problems with the map, it might be simpler to use the WriteBatch object directly instead of trying to use the map first.

Reply:
The reason the insertBatch map is being used to is force sort the keys.  RocksDB indicated sorting would be faster for doing an insert.  The comment on line 54 was trying to imply this information.  Let me know if I can be clearer.